common/core/desktop/src/kmx/kmx_processor.cpp

```diff
@@ -194,3 +194,40 @@ constexpr km_kbp_attr const engine_attrs = {
 km_kbp_attr const & kmx_processor::attributes() const {
   return engine_attrs;
 }
+
+km_kbp_keyboard_key * kmx_processor::get_key_list() const  {
+  // Iterate through the groups and get the rules with virtual keys
+  // and store the key along with the modifer.
+  const uint32_t group_cnt = _kmx.GetKeyboard()->Keyboard->cxGroupArray;
+  const LPGROUP group_array = _kmx.GetKeyboard()->Keyboard->dpGroupArray;
+  uint16_t vk_count = 0;
+  GROUP *p_group;
+
+  for(auto i = decltype(group_cnt){0}; i < group_cnt; i++)
+  {
+    if(group_array[i].fUsingKeys)
+    {
+      vk_count += group_array[i].cxKeyArray;
+    }
+  }
+
+  km_kbp_keyboard_key *rules = new km_kbp_keyboard_key[vk_count + 1];
+  int n = 0;
+  for(auto i = decltype(group_cnt){0}; i < group_cnt; i++)
+  {
+    p_group = &group_array[i];
+    if(p_group->fUsingKeys)
+    {
+      for(auto j = decltype(p_group->cxKeyArray){0}; j < p_group->cxKeyArray; j++)
+      {
+        // If we have a key rule for the key add it to the list
+        rules[n].key = p_group->dpKeyArray[j].Key;
+        rules[n].modifier_flag = p_group->dpKeyArray[j].ShiftFlags;
+        n++;
+      }
+    }
+  }
```

Comment:
This may return the same key combination multiple times from multiple groups. I think that may be a bit ugly -- we probably need to make sure that they are not repeated. Might be best to build that with a hash table (scanning for repeats naively would be `O(n*n)`)?

We do have a limited set of possible virtual key codes -- at least until we support touch layouts in the future which can define arbitrary virtual keys (codes > 256).